mlflow/models/signature.py

```diff
@@ -0,0 +1,464 @@
+
+    @classmethod
+    def from_json(cls, json_str: str):
+        return cls([ColSpec(**x) for x in json.loads(json_str)])
+
+
+class ModelSignature(object):
+    """
+    ModelSignature specifies schema of model's inputs and outputs.
+
+    The current supported schema for both the input and the output is a data-frame like schema
+    defined as a list of column specification :py:class:`ColSpec`. Columns can be named and must
+    specify their data type. Currently the list of supported types is limited to scalar data types
+    defined in :py:class:`DataType` enum.
+
+    ModelSignature can be inferred from training dataset and model predictions using
+    :py:func:`mlflow.models.signature.infer_signature`, or alternatively constructed by hand by
+    passing a lists of input and output column specifications.
+    """
+
+    def __init__(self, inputs: Schema, outputs: Schema = None):
+        if not isinstance(inputs, Schema):
+            raise TypeError("inputs must be mlflow.models.signature.Schema, got '{}'".format(
+                type(inputs)))
+        if outputs is not None and not isinstance(outputs, Schema):
+            raise TypeError("outputs must be either None or mlflow.models.signature.Schema, "
+                            "got '{}'".format(type(inputs)))
+        self.inputs = inputs
+        self.outputs = outputs
+
+    def to_dict(self) -> Dict[str, str]:
+        """
+        Serialize into a 'jsonable' dictionary.
+
+        Input and output schema are represented as json strings. This is so that the
+        representation is compact when embedded in a MLmofel yaml file.
+
+        :return: dictionary representation with input and output shcema represented as json strings.
+        """
+
+        return {
+            "inputs": self.inputs.to_json(),
+            "outputs": self.outputs.to_json() if self.outputs is not None else None
+        }
+
+    @classmethod
+    def from_dict(cls, signature_dict: Dict[str, str]):
+        """
+        Deserialize from dictionary representation.
+
+        :param signature_dict: Dictionary representation of model signature.
+                               Expected dictionary format:
+                               `{'inputs': <json string>, 'outputs': <json string>" }`
+
+        :return: ModelSignature populated with the data form the dictionary.
+        """
+        inputs = Schema.from_json(signature_dict["inputs"])
+        if "outputs" in signature_dict and signature_dict["outputs"] is not None:
+            outputs = Schema.from_json(signature_dict["outputs"])
+            return cls(inputs, outputs)
+        else:
+            return cls(inputs)
+
+    def __eq__(self, other) -> bool:
+        return self.inputs == other.inputs and self.outputs == other.outputs
+
+    def __repr__(self) -> str:
+        return json.dumps({"ModelSignature": self.to_dict()}, indent=2)
+
+
+ModelInputExample = TypeVar('ModelInputExample', pd.DataFrame, np.ndarray, dict, list)
```

Comment:
Neat: `TypeError: Unexpected type of input_example. Expected one of (pandas.DataFrame, numpy.ndarray, dict, list), got <class 'int'>`